### PR TITLE
add DnsConfig

### PR DESCRIPTION
**Pull request by glwlg**
add DnsConfig in ContainerSpec

### CODE REVIEW

**Comment by glwlg** [comment]
@johnflavin when I update a docker service for change replicas 3 to replicas 5 , it will stop the old container, because there is no DNSConfig
snipaste20170413_104009
snipaste20170413_104716
snipaste20170413_104737
snipaste20170413_104822
snipaste20170413_104938

sorry for my english

**Comment by johnflavin** [comment]
I'll add a comment to a line in the code to clarify my meaning.

File: src/main/java/com/spotify/docker/client/messages/swarm/ContainerSpec.java [hunk 1]
```diff
@@ -326,6 +331,17 @@ public Builder withStopGracePeriod(final long stopGracePeriod) {
       return this;
     }
 
+    public abstract Builder dnsConfig(DnsConfig dnsConfig);
+
+    /**
+     * @deprecated  As of release 7.0.0, replaced by {@link #dnsConfig(DnsConfig)}.
+     */
+    @Deprecated
+    public Builder withDnsConfig(final DnsConfig dnsConfig) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/ContainerSpec.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    public abstract Builder nameServers(String... nameServers);
+
+    public abstract Builder nameServers(List<String> nameServers);
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #nameServers(String...)}.
+    */
+    @Deprecated
+    public Builder withNameServers(final String... nameServers) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        nameServers(nameServers);
+      }
+      return this;
+    }
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #nameServers(List)}.
+    */
+    @Deprecated
+    public Builder withNameServers(final List<String> nameServers) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    
+    public abstract Builder search(String... search);
+
+    public abstract Builder search(List<String> search);
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #search(String...)}.
+    */
+    @Deprecated
+    public Builder withsearch(final String... search) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        search(search);
+      }
+      return this;
+    }
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #search(List)}.
+    */
+    @Deprecated
+    public Builder withsearch(final List<String> search) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    
+    public abstract Builder options(String... options);
+
+    public abstract Builder options(List<String> options);
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #options(String...)}.
+    */
+    @Deprecated
+    public Builder withoptions(final String... options) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        options(options);
+      }
+      return this;
+    }
+
+    /**
+    * @deprecated  As of release 7.0.0, replaced by {@link #options(List)}.
+    */
+    @Deprecated
+    public Builder withoptions(final List<String> options) {
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: Please remove this method.

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      options(options);
+      return this;
+    }
+
+    public abstract com.spotify.docker.client.messages.swarm.DnsConfig build();
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: You should not need to give the full package of the class here. Just the class name will do. 

(Your IDE may have inserted this for you. I know mine did that for a while, but I don't remember how I was able to get it to stop doing that.)

**Comment by glwlg** [comment]
sorry.
thank you for the explanation

File: src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.List;
+import javax.annotation.Nullable;
+
+@AutoValue
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public abstract class DnsConfig {
+
+  @Nullable
+  @JsonProperty("Nameservers")
+  public abstract ImmutableList<String> nameServers();
```
Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: are each of the 3 fields in the DNSConfig element actually nullable? I would imagine that having a DNSConfig without the list of nameservers is not allowed

Comment (glwlg) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: snipaste20170420_095719
yes, I think these are nullable. In the picture , it's a service. Sorry about Deprecated , thank reply.

Comment (glwlg) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: My Docker version is 17.03.1-ce, and I create this service with `docker service create \
--replicas 1 --name nginx --publish 80:80 nginx` .

Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/messages/swarm/DnsConfig.java]: thanks for clarifying

**Pull request merged**